codegen/lib/src/type.dart

```diff
@@ -111,6 +118,146 @@ class MessageType {
+        'false',
+        'final',
+        'finally',
+        'for',
+        'if',
+        'implements',
+        'in',
+        'is',
+        'new',
+        'null',
+        'rethrow',
+        'return',
+        'super',
+        'switch',
+        'this',
+        'throw',
+        'true',
+        'try',
+        'var',
+        'void',
+        'while',
+        'with',
+        'yield',
+    ];
+
+    final MessageType declaringType;
+    final FieldDescriptorProto descriptor;
+
+    FieldDeclaration(this.declaringType, this.descriptor);
+
+    /// The name of the field as declared in Protobuf.
+    String get protoName => descriptor.name;
+
+    /// The name of the field "translated directly" to Dart, i.e. in "javaCase".
+    ///
+    /// See [escapedDartName] for code generation.
+    ///
+    String get dartName {
+        var protoName = descriptor.name;
+        var words = protoName.split('_');
+        var first = words[0];
+        var capitalized = List.of(words.map(_capitalize));
+        capitalized[0] = first;
+        return capitalized.join('');
+    }
+
+    static String _capitalize(String word) {
+        return word.isEmpty
+            ? word
+            : '${word[0].toUpperCase()}${word.substring(1)}';
+    }
+
+    /// A field name which is safe to use in Dart.
+    ///
+    /// The escaping mechanism is the same as the one provided by the Dart Protoc plugin.
+    ///
+    /// If a Proto message field has a name which clashes with a Dart keyword, a suffix is added to
+    /// the field. The suffix consists of an underscore (`_`) and the field number. For example,
+    /// given a message:
+    /// ```proto
+    /// message Glue {
+    ///     bool super = 1;
+    /// }
+    /// ```
+    /// The escaped name of the field, which will be used for code generation, is `Glue.super_1`.
+    ///
+    String get escapedDartName {
+        var name = dartName;
+        if (_DART_KEYWORDS.contains(name)) {
+            return '${name}_${descriptor.number}';
+        } else {
+            return name;
+        }
+    }
+
+    /// Checks if this field is `repeated`.
+    bool get isRepeated =>
+        descriptor.label == FieldDescriptorProto_Label.LABEL_REPEATED;
+
+    /// Checks if this field is a `map`.
+    bool get isMap {
+        if (!isRepeated) {
+            return false;
+        }
+        if (descriptor.type != FieldDescriptorProto_Type.TYPE_MESSAGE) {
+            return false;
+        }
+        var mapEntryTypeName = _capitalize(dartName) + 'Entry';
+        return _simpleName(descriptor.typeName) == mapEntryTypeName;
+    }
+
+    String _simpleName(String fullName) {
+        var start = fullName.lastIndexOf('.') + 1;
+        return fullName.substring(start);
+    }
+
+    /// Finds the given custom option on the field.
+    ///
+    /// Returns the value of the field or `Optional.empty()` if the option is not defined
+    /// on the field.
+    ///
+    Optional<T> getOption<T>(Extension<T> option) {
```

Comment:
Not sure how common it's to have a getter-style naming in Dart, but in any case I would go with `findOption()`. 

`getSmth()` feels more like an endpoint which throws an exception if that `smth` is missing.